Make icon flashing configurable

ferdnyc opened the pull request:
I've really found this extension useful, after the loss of the blue-dot feature. Thanks so much for creating it.

Personally, though, I find the flashing to be a little distracting, and feel a static "notification waiting" icon is sufficient for my needs. So, I added a preference to select how the icon is flashed on activation. Currently there are only two options ("enable" and "disable", to switch flashing on/off completely), but in theory the choices could be expanded to allow for control over the speed of the flashing, a certain number of flashes before stopping, etc.

As of this PR it's not _quite_ perfect, because when you change the setting, any existing icons aren't affected — if you disable flashing, any existing flashing icons will continue to flash. And if you enable flashing, only new icons shown after that point will flash. That hasn't bothered me since I only changed the setting once, but I recognize that it could be improved on.

It's certainly possible for `options.js` to update any existing icons whenever the `flashIcon` setting is changed. If I get inspired, or if you think it's necessary for this patch to merge (assuming you're interested at all), I may still go back and add that. It would also be a good opportunity to add flashing to the Preview icon in the options window (when enabled), so that it accurately previewed the current settings.

ferdnyc committed:
Add flashIcon setting to options

ferdnyc committed:
Support pref to disable icon flashing

willg101 commented on bg.js:
```diff
@@ -53,23 +53,41 @@ chrome.tabs.onUpdated.addListener( ( id, changeInfo, changedTab ) =>
-					}, ${FLASH_INTERVAL_MS} );
-				} )` 
-			} );
+			if (settings.flashIcon == 'enable' )
```
I think to avoid code duplication it would be better to put this `if` statement right above the code that triggers the flashing, e.g.,:
```js
if (settings.flashIcon == 'enable')
{
	let showingOld = false;
	el.interval = setInterval( () =>
	{
		el.href = showingOld ? "${alertIcon}" : el.oldHref;
		showingOld = !showingOld;
	}, ${FLASH_INTERVAL_MS} );
}
```

ferdnyc replied:
I agree that definitely makes sense. I struggled with that a bit initially, attempting to do the alert icon setup in two separate `executeScript()` runs (one always performed, the second conditional on `settings.flashIcon`), but that didn't really work out. 

That was before I changed the semantics of `settings.flashIcon` to having the value "enabled" or "disabled", though. With those string values, I should be able to include a conditional inside the script code, the same way `alertIcon` is used, basically:

```js
chrome.tabs.executeScript( id, {
  code: `document.querySelectorAll( 'link[rel*="icon"]' ).forEach( el =>
    {
      if ( el.oldHref )
      {
        return;
      }
      el.oldHref = el.href;
      el.href = "${alertIcon}"
      if ("${settings.flashIcon}" == 'enabled') {
        let showingOld = false;
        el.interval = setInterval( () =>
        {
          el.href = showingOld ? "${alertIcon}" : el.oldHref;
          showingOld = !showingOld;
        }, ${FLASH_INTERVAL_MS} )
      }
    } );`
```

I'll add a commit with that change (or something equivalent) later tonight.

ferdnyc replied:
The most recent commit should address this.

willg101 commented:
Cool - thank you for contributing this feature!

> As of this PR it's not quite perfect, because when you change the setting, any existing icons aren't affected — if you disable flashing, any existing flashing icons will continue to flash. And if you enable flashing, only new icons shown after that point will flash. That hasn't bothered me since I only changed the setting once, but I recognize that it could be improved on.

That does not bother me either; I imagine users would understand that the setting will only take effect going forward.

> [...] It would also be a good opportunity to add flashing to the Preview icon in the options window (when enabled), so that it accurately previewed the current settings.

That would be really nice, but only if you have time. Lack of flashing/static icons in the preview would not prevent me from merging. 😄

ferdnyc committed:
Move conditional into executeScript() code

I also switched the semantics to `flashIcon != 'disable'`, so that 
`flashIcon` could potentially support multiple enabled states ("fast" / 
"slow" / etc.).
